### PR TITLE
types: fix BrowserWindow mock type

### LINKED ISSUE #817: BrowserWindow mock type
**Describe the bug**

The types of Mock of BrowserWindow use `vitest` internal type `MaybeMocked`

https://github.com/cawa-93/vite-electron-builder/blob/1daa1d55c69423ba515045c1f4763c462a0914fd/packages/main/tests/unit.spec.ts#L1-L24

**Fix**

Use `MockedObject` will fix this issue but the type of `bw.prototype` is `any`.

```ts
import type {MockedObject} from 'vitest';
import {beforeEach, expect, test, vi} from 'vitest';
import {restoreOrCreateWindow} from '../src/mainWindow';

import {BrowserWindow} from 'electron';

/**
 * Mock real electron BrowserWindow API
 */
vi.mock('electron', () => {

  const bw = vi.fn() as MockedObject<typeof BrowserWindow>;
  bw.getAllWindows = vi.fn(() => bw.mock.instances);
  // bw.prototype is any
  bw.prototype.loadURL = vi.fn();
  bw.prototype.on = vi.fn();
  bw.prototype.destroy = vi.fn();
  bw.prototype.isDestroyed = vi.fn();
  bw.prototype.isMinimized = vi.fn();
  bw.prototype.focus = vi.fn();
  bw.prototype.restore = vi.fn();

  return {BrowserWindow: bw};
});
```

For the strictest typing, use `MockedClass` which ports the prototype part of `BrowerWindow` class. Unfortunately there's a bug of `MockedClass` type in `vistest` (see https://github.com/vitest-dev/vitest/issues/1730 , basically they put args and returns in wrong positions lol). I have to include a manual fix for that.

```ts
import type {Constructable, Mocked, MockInstance} from 'vitest';
import {beforeEach, expect, test, vi} from 'vitest';
import {restoreOrCreateWindow} from '../src/mainWindow';

import {BrowserWindow} from 'electron';

// Manual fix of MockedClass type
// See https://github.com/vitest-dev/vitest/issues/1730
type MockedClass<T extends Constructable> = MockInstance<T extends new (...args: infer P) => any ? P : never, InstanceType<T>> & {
  prototype: T extends {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      prototype: any;
  } ? Mocked<T['prototype']> : never;
} & T;

/**
 * Mock real electron BrowserWindow API
 */
vi.mock('electron', () => {

  // Use "as unknown as" because vi.fn() does not have static methods
  const bw = vi.fn() as unknown as MockedClass<typeof BrowserWindow>;
  bw.getAllWindows = vi.fn(() => bw.mock.instances);
  bw.prototype.loadURL = vi.fn();
  // Use "any" because the on function is overloaded
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  bw.prototype.on = vi.fn<any>();
  bw.prototype.destroy = vi.fn();
  bw.prototype.isDestroyed = vi.fn();
  bw.prototype.isMinimized = vi.fn();
  bw.prototype.focus = vi.fn();
  bw.prototype.restore = vi.fn();

  return {BrowserWindow: bw};
});
```

All the fix does not change the runtime behaviors, just types.

### PR DESCRIPTION
**Pull request by Tanimodori**
Fix #817

### CODE REVIEW

File: packages/main/tests/unit.spec.ts [hunk 1]
```diff
@@ -1,19 +1,30 @@
-import type {MaybeMocked} from 'vitest';
+import type {Constructable, Mocked, MockInstance} from 'vitest';
 import {beforeEach, expect, test, vi} from 'vitest';
 import {restoreOrCreateWindow} from '../src/mainWindow';
 
 import {BrowserWindow} from 'electron';
 
+// Manual fix of MockedClass type
+// See https://github.com/vitest-dev/vitest/issues/1730
```
Comment (cawa-93) [inline packages/main/tests/unit.spec.ts:8]: Pls, use here JSDoc instead

File: packages/main/tests/unit.spec.ts [hunk 2]
```diff
@@ -1,19 +1,30 @@
...
+type MockedClass<T extends Constructable> = MockInstance<T extends new (...args: infer P) => any ? P : never, InstanceType<T>> & {
```
Comment (cawa-93) [inline packages/main/tests/unit.spec.ts:9]: Disable eslink warn here to

**Review by cawa-93 — CHANGES_REQUESTED** [review]

File: packages/main/tests/unit.spec.ts (hunk 1 above)
Comment (Tanimodori) [reply packages/main/tests/unit.spec.ts]: fixed

File: packages/main/tests/unit.spec.ts (hunk 2 above)
Comment (Tanimodori) [reply packages/main/tests/unit.spec.ts:12]: fixed

**Pull request merged**